internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,215 @@
+	select {
+	case <-sr.doneCh: // will non-blockingly receive whenever doneCh would be closed
+		sr.log.Info("status reporter is already stopped")
+		return nil
+	case sr.stopperCh <- true:
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("failed to stop the status reporter: %w", ctx.Err())
+	}
+}
+
+func (sr *StatusReporter) SetConditions(ctx context.Context, conditions []metav1.Condition) error {
+	// immediately register a new heartbeat upon receive one from the client/tenant side
+	addonInstance := &addonsv1alpha1.AddonInstance{}
+	if err := sr.addonInstanceInteractor.GetAddonInstance(ctx, types.NamespacedName{Name: "addon-instance", Namespace: sr.addonTargetNamespace}, addonInstance); err != nil {
+		return fmt.Errorf("failed to get the AddonInstance: %w", err)
+	}
+
+	// making a deep-copy for current Conditions for rolling back in case of failures
+	previousConditions := addonInstance.Status.Conditions
+
+	newConditions := addonInstance.Status.Conditions
+	for _, condition := range conditions {
+		meta.SetStatusCondition(&newConditions, condition)
+	}
+	addonInstance.Status.Conditions = newConditions
+	addonInstance.Status.ObservedGeneration = addonInstance.Generation
+	addonInstance.Status.LastHeartbeatTime = metav1.Now()
+
+	if err := sr.addonInstanceInteractor.UpdateAddonInstanceStatus(ctx, addonInstance); err != nil {
+		return fmt.Errorf("failed to update AddonInstance Status: %w", err)
+	}
+
+	// rollbackAddonInstanceStatusUpdate() will be called when the `conditions` would fail to be sent on the `sr.updateCh` channel
+	rollbackAddonInstanceStatusUpdate := func() error {
+		addonInstance.Status.Conditions = previousConditions
+		addonInstance.Status.LastHeartbeatTime = metav1.Now()
+
+		if err := sr.addonInstanceInteractor.UpdateAddonInstanceStatus(ctx, addonInstance); err != nil {
+			return fmt.Errorf("failed to update AddonInstance Status: %w", err)
+		}
+		return nil
+	}
+
+	select {
+	case <-sr.doneCh:
+		sr.log.Info("StatusReporter found to be stopped")
+		return nil
+	case sr.updateCh <- updateOptions{conditions: &conditions}: // near-instantly received by the StatusReporter loop
+		return nil
+	case <-ctx.Done():
+		sr.log.Error(ctx.Err(), "failed to send the heartbeat's Conditions to the StatusReporter")
+		sr.log.Info("rolling back the AddonInstance Status update...")
+		// not 100% full-proof rollback because in theory, the rollback could still fail.
+		// Yet in that case too, eventual consistency would still be preserved because our AddonInstance.Status.Conditions would get re-synchronized to `sr.latestConditions` in the next iteration of the StatusReporter loop.
+		if err := rollbackAddonInstanceStatusUpdate(); err != nil {
+			return fmt.Errorf("failed to rollback the AddonInstance Status update: %w", err)
+		}
+		return nil
+	}
+}
+
+func (sr *StatusReporter) ReportAddonInstanceSpecChange(ctx context.Context, newAddonInstance addonsv1alpha1.AddonInstance) error {
```

Comment:
Any specific reason for doing so?
This method isn't meant to modify the incoming AddonInstance object. All it is meant to do is to just pass its (input AddonInstance object's) info to StatusReporter's Heartbeat Reporter loop, which it is doing :/

Reply:
It's not about modifying the `AddonInstance` object, because you can still modify the struct when any property is a pointer:
https://go.dev/play/p/LPLv59t5E5n

What you are doing here is to make a incomplete copy of the struct, which can easily lead to bugs.
If you want to make a complete copy, you have to call the generated `DeepCopy` functions, but IMHO that only allocates extra memory without any benefit.